File: server/src/main/com/broll/gainea/server/core/goals/impl/e2/G_Snow.java
package com.broll.gainea.server.core.goals.impl.e2;

import com.broll.gainea.server.core.bot.strategy.GoalStrategy;
import com.broll.gainea.server.core.goals.CustomOccupyGoal;
import com.broll.gainea.server.core.goals.GoalDifficulty;
import com.broll.gainea.server.core.map.AreaType;
import com.broll.gainea.server.core.map.ExpansionType;
import com.broll.gainea.server.core.utils.LocationUtils;

import java.util.stream.Collectors;

public class G_Snow extends CustomOccupyGoal {

    private final static int COUNT = 5;

    public G_Snow() {
        super(GoalDifficulty.EASY, "Erobere " + COUNT + " Schneegebiete");
        setExpansionRestriction(ExpansionType.ICELANDS);
        setProgressionGoal(COUNT);
    }

    @Override
    public void check() {
        int count = (int) LocationUtils.getControlledLocationsIn(player, ExpansionType.ICELANDS).stream().filter(it -> LocationUtils.isAreaType(it, AreaType.SNOW)).count();
        updateProgression(count);
        if (count >= COUNT) {
            success();
        }
    }

    @Override
    public void botStrategy(GoalStrategy strategy) {
        strategy.setRequiredUnits(COUNT);
        strategy.updateTargets(game.getMap().getAllAreas().stream().filter(it -> it.getType() == AreaType.SNOW).collect(Collectors.toSet()));
    }
}

File: server/src/main/com/broll/gainea/server/core/goals/impl/e2/G_Snow.kt
package com.broll.gainea.server.core.goals.impl.e2

import com.broll.gainea.server.core.bot.strategy.GoalStrategy
import com.broll.gainea.server.core.goals.CustomOccupyGoal
import com.broll.gainea.server.core.goals.GoalDifficulty
import com.broll.gainea.server.core.map.AreaType
import com.broll.gainea.server.core.map.ExpansionType
import com.broll.gainea.server.core.utils.getControlledLocationsIn
import com.broll.gainea.server.core.utils.isAreaType

class G_Snow : CustomOccupyGoal(GoalDifficulty.EASY, "Erobere " + COUNT + " Schneegebiete") {
    init {
        setExpansionRestriction(ExpansionType.ICELANDS)
        progressionGoal = COUNT
    }

    override fun check() {
        val count = player.getControlledLocationsIn(ExpansionType.ICELANDS)
                .count { it.isAreaType(AreaType.SNOW) }
        updateProgression(count)
        if (count >= COUNT) {
            success()
        }
    }

    override fun botStrategy(strategy: GoalStrategy) {
        strategy.setRequiredUnits(COUNT)
        strategy.updateTargets(game.map.allAreas.filter { it.type == AreaType.SNOW }.toSet())
    }

    companion object {
        private const val COUNT = 5
    }
}
